akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala

```diff
@@ -610,8 +610,16 @@ final class Balance[T](val outputPorts: Int, waitForAllDownstreams: Boolean) ext
 
     private def dequeueAndDispatch(): Unit = {
       val out = pendingQueue.dequeue()
-      push(out, grab(in))
-      if (!noPending) pull(in)
+      if (out ne null) {
```

Comment:
add a comment, that this can happen because of the reentrancy in line L620 (we thrown out an "out" and there is no new one)